manticore/core/smtlib/constraints.py

```diff
@@ -146,7 +146,7 @@ def __get_related(self, related_to=None):
                             break
 
                     variables = get_variables(constraint)
-                    if related_variables & variables:
+                    if related_variables & variables or not variables:
```

Comment:
With this added "this" test passes with both get_related enabled and disabled

Reply:
As long as both are passing, let's set the default back to `True`

Reply:
Well I'm not 100% This single testcase representes all possible fails explained in issue https://github.com/trailofbits/manticore/issues/1679

Reply:
Devils advocate, here...  What if `related_to` is `A` and `constraint` is something like `B && !B`, i.e., `constraint` is unsatisfiable and has variables, but they do not overlap with those of `related_to`?

Reply:
Yes get_related should work only on previously feasible path contraints.
If you have feasible `A & B & C` and want to check if `B \/ C` can be True then you do not need to bother the solver with the A.

Reply:
> get_related should work only on previously feasible path contraints.
In other words, you consider it a precondition that the constraint set on which `__get_related` is called is feasible?

Reply:
Yes. I know it is not enforced.
We should make it a pre-condition or attempt to solve the full constraint set and trust that some cache magic prevents it to do the solving over the whole thing again and again.

Reply:
> Yes. I know it is not enforced.
> We should make it a pre-condition or attempt to solve the full constraint set and trust that some cache magic prevents it to do the solving over the whole thing again and again.

I don't know how reasonable this approach is. `__get_related` is called by `to_string`, and `to_string` is something that one might reasonably call to debug a bad constraint set.

Moreover, `to_string` is called by `can_be_true`:
https://github.com/trailofbits/manticore/blob/2c90ee07c9e691304db97f573bff61ee220ffbb6/manticore/core/smtlib/solver.py#L433
So there's a little bit of a chicken-and-egg problem here.

If it were up to me, I would not insist that `__get_related` be called on only satisfiable constraint sets.

Reply:
Many constraints do not overlap in terms of the memory they reference. Constraint independence divides constraint sets into disjoint independent subsets based on the symbolic variables and memory cells they reference. By explicitly tracking these subsets, it can frequently eliminate irrelevant constraints prior to sending it to the solver.
For example, given the constraint set {i < j, j < 20, k > 0}, a query of whether i = 20 just requires the first two constraints.  This was first explained in [EXE] https://web.stanford.edu/~engler/exe-ccs-06.pdf

We can move the optimizations/transformations that are controlled via kwargs in the `to_string()` to a different `ConstraintSet` method. Like `slice_related_to(..)` and `make_bindings` so it is done more explicit.

If the problem is to print a ConstraintSet for debugging you can just just do not slice the ConstraintSet.

``` print (str(cs)) ```

There is also a forgotten `pretty_print` for expressions.

Reply:
Fwiw, I use the `pretty_print` all the time while debugging

Reply:
Moved the disabled get_related as a special method of a ConstraintSet to slice it when the user wants it.
Also added "some" tests for it.